File: algorithm/3월4주차/현욱은 괄호왕이야!!_최철원.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Main {
	public static void main (String[] args) throws java.lang.Exception {
	    BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	    int num = Integer.parseInt(br.readLine());
	    String str = br.readLine();
	    Stack<Integer> s = new Stack<>();
	    s.add(-1);
	    int answer = 0;
	    for(int i=0;i<num;i++){
	        char c = str.charAt(i);
	        if(c=='(') s.add(i);
	        else{
	            s.pop();
	            if(s.empty()){
	                s.add(i);
	                continue;
	            }
	            int idx = s.peek();
	            answer = Math.max(answer,i-idx);
	        }
	    }
		System.out.print(answer);
	}
}

File: algorithm/3월4주차/현욱은 괄호왕이야!!_최철원.cpp
#include <iostream>
#include <string>
#include <algorithm>
#include <stack>
using namespace std;
struct info{
    char c;
    int idx;
};

int main() {
    ios_base::sync_with_stdio(false); cin.tie(NULL); cout.tie(NULL);
    int num,answer=0;
    string str;
    cin>>num>>str;
    stack<info> s;
    
    for(int i=0;i<num;i++){
        char c = str[i];
        if(s.empty()){
            s.push({c,i});
            continue;
        }
        char topChar = s.top().c;
        if(topChar=='(' && c==')') s.pop();
        else s.push({c,i});
    }
    int temp = num;
    while(!s.empty()){
        int idx = s.top().idx;
        int diff = temp-idx-1;
        if(diff>1) answer = max(answer, diff);
        temp=idx;
        s.pop();
    }
    answer = max(answer,temp);
    cout<<answer;
    return 0;
}
